src/CLIOptions.cpp

```diff
@@ -0,0 +1,73 @@
+#include "CLIOptions.h"
+#include "MainOpt.h"
+#include "uri.h"
+#include <CLI/CLI.hpp>
+
+CLI::Option *uri_option(CLI::App &App, std::string Name, uri::URI &URIArg,
+                        CLI::callback_t Fun, std::string Description,
+                        bool Defaulted) {
+
+  CLI::Option *Opt = App.add_option(Name, Fun, Description, Defaulted);
+  Opt->set_custom_option("URI", 1);
+  if (Defaulted) {
+    Opt->set_default_str(URIArg.getURIString());
+  }
+  return Opt;
+}
+
+/// Use for adding a URI option
+CLI::Option *add_option(CLI::App &App, std::string Name, uri::URI &URIArg,
+                        std::string Description = "", bool Defaulted = false) {
+  CLI::callback_t Fun = [&URIArg](CLI::results_t Results) {
+    URIArg.parse(Results[0]);
+    return true;
+  };
+
+  return uri_option(App, Name, URIArg, Fun, Description, Defaulted);
+}
+
+/// Use for adding a URI option, if the URI is given then TrueIfOptionGiven is
+/// set to true
+CLI::Option *add_option(CLI::App &App, std::string Name, uri::URI &URIArg,
+                        bool &TrueIfOptionGiven, std::string Description = "",
+                        bool Defaulted = false) {
+  CLI::callback_t Fun = [&URIArg, &TrueIfOptionGiven](CLI::results_t Results) {
+    TrueIfOptionGiven = true;
+    URIArg.parse(Results[0]);
+    return true;
+  };
+
+  return uri_option(App, Name, URIArg, Fun, Description, Defaulted);
+}
+
+void setCLIOptions(CLI::App &App, MainOpt &MainOptions) {
+  // disable ini config file
+  App.set_config();
+  // and add option for json config file instead
+  App.add_option("--config-file", MainOptions.config_filename,
+                 "Specify a json file to set config")
+      ->check(CLI::ExistingFile);
+
+  add_option(
```

Comment:
Is there any difference between ``App.add_option("...)`` and ``add_option(App,...`` ? If not, can we use only one of the two (to keep the style consistent)?

Reply:
The latter is implemented by me to allow adding `URI` options. It is not really an overload as the former is a method of `App` but the suggestion in the CLI11 docs is to use `add_option` for your own types.